pkg/trace/api/api.go

```diff
@@ -263,30 +268,50 @@ func (r *HTTPReceiver) replyTraces(v Version, w http.ResponseWriter) {
 	}
 }
 
+func getTraceCount(req *http.Request) (traceCount int64) {
+	if traceCountStr := req.Header.Get(TraceCountHeader); traceCountStr != "" {
+		var err error
+		traceCount, err = strconv.ParseInt(traceCountStr, 10, 64)
+		if err != nil {
+			log.Errorf("unable to parse HTTP header %s: %s", TraceCountHeader, traceCountStr)
+		}
+	}
+	return traceCount
+}
+
 // handleTraces knows how to handle a bunch of traces
 func (r *HTTPReceiver) handleTraces(v Version, w http.ResponseWriter, req *http.Request) {
-	if !r.PreSampler.Sample(req) {
+	traceCount := getTraceCount(req)
+
+	if !r.PreSampler.SampleWithCount(traceCount) {
 		io.Copy(ioutil.Discard, req.Body)
 		w.WriteHeader(r.presamplerResponse)
 		r.replyTraces(v, w)
 		metrics.Count("datadog.trace_agent.receiver.payload_refused", 1, nil, 1)
 		return
 	}
 
-	traces, ok := getTraces(v, w, req)
-	if !ok {
-		return
-	}
-
-	r.replyTraces(v, w)
-
 	ts := r.Stats.GetTagStats(info.Tags{
 		Lang:          req.Header.Get("Datadog-Meta-Lang"),
 		LangVersion:   req.Header.Get("Datadog-Meta-Lang-Version"),
 		Interpreter:   req.Header.Get("Datadog-Meta-Lang-Interpreter"),
 		TracerVersion: req.Header.Get("Datadog-Meta-Tracer-Version"),
 	})
 
+	traces, err := getTraces(v, w, req)
+	if err != nil {
+		if strings.HasPrefix(err.Error(), "Cannot decode") {
```

Comment:
This is not how we check errors. If you really want to change `getTraces` to return errors, you'll need to create a set of custom errors. That's the idiomatic way to do it in Go, but it will be tedious. I'll try to illustrate. 
```go
type decodingError struct {err error}

func (e *decodingError) Error() string { return fmt.Sprintf("decoding error: %s", e.err) }

type formatError struct {t string}

func (e *formatError) Error() string { return fmt.Sprintf("unsupported media type: %s", e.t) }

type versionError struct {v Version}

func (e *versionError) Error() string { return fmt.Sprintf("unsupported version: %s", e.v) }
```
Then, you would here do:
```go
traces, err := getTraces(v, req)
switch e := err.(type) {
    case decodingError:
        if traceCount > 0 {
            atomic.AddInt64(&ts.TracesDropped.DecodingError, traceCount)
        } else {
            metrics.Count("datadog.trace_agent.receiver.decoding_error", 1, nil, 1)
        }
        log.Errorf("Cannot decode %s traces payload: %v", v, err)
        httpDecodingError(err, []string{tagTraceHandler, fmt.Sprintf("v:%s", v)}, w)

    case formatError:
        httpFormatError(w, v, fmt.Errorf("unsupported media type: %q", mediaType))

    case versionError:
        httpEndpointNotSupported([]string{tagTraceHandler, fmt.Sprintf("v:%s", v)}, w)
}
```
This would significantly clean up `getTraces` to the extent that it's equally long as the above switch statement, making this whole suggestion overkill, even if more correct. What I would like to suggest instead is to either:
* Inline all of `getTraces` into `handleTraces` right here. It doesn't make much sense to have it as a separate function to be honest, or,
* Just use the change I initially suggested [here](https://github.com/DataDog/datadog-agent/pull/3738#issuecomment-506373265) and nothing else.

I am also ok with the overkill example above, but frankly I think it's a bit too much. The bottom line is that we shouldn't mix returning errors with handling or logging those very same errors. It also seems weird that a function called `getTraces` is replying to the HTTP client. Personally I think it'd serve as a nice cleanup to inline `getTraces` into `handleTraces`

Reply:
I tried it out:
```diff
diff --git a/pkg/trace/api/api.go b/pkg/trace/api/api.go
index aeb02b13..06235a82 100644
--- a/pkg/trace/api/api.go
+++ b/pkg/trace/api/api.go
@@ -273,8 +273,32 @@ func (r *HTTPReceiver) handleTraces(v Version, w http.ResponseWriter, req *http.
 		return
 	}
 
-	traces, ok := getTraces(v, w, req)
-	if !ok {
+	var traces pb.Traces
+	mediaType := getMediaType(req)
+	switch v {
+	case v01:
+		switch mediaType {
+		case "application/json", "text/json", "":
+			// ok
+		default:
+			httpFormatError(w, v, fmt.Errorf("unsupported media type: %q", mediaType))
+			return
+		}
+		var spans []pb.Span
+		if err := json.NewDecoder(req.Body).Decode(&spans); err != nil {
+			log.Errorf("Cannot decode %s traces payload: %v", v, err)
+			httpDecodingError(err, []string{tagTraceHandler, fmt.Sprintf("v:%s", v)}, w)
+			return
+		}
+		traces = tracesFromSpans(spans)
+	case v02, v03, v04:
+		if err := decodeReceiverPayload(req.Body, &traces, v, mediaType); err != nil {
+			log.Errorf("Cannot decode %s traces payload: %v", v, err)
+			httpDecodingError(err, []string{tagTraceHandler, fmt.Sprintf("v:%s", v)}, w)
+			return
+		}
+	default:
+		httpEndpointNotSupported([]string{tagTraceHandler, fmt.Sprintf("v:%s", v)}, w)
 		return
 	}
 
@@ -479,48 +503,6 @@ func (r *HTTPReceiver) Languages() string {
 	return strings.Join(str, "|")
 }
 
-func getTraces(v Version, w http.ResponseWriter, req *http.Request) (pb.Traces, bool) {
-	var traces pb.Traces
-	mediaType := getMediaType(req)
-	switch v {
-	case v01:
-		// We cannot use decodeReceiverPayload because []model.Span does not
-		// implement msgp.Decodable. This hack can be removed once we
-		// drop v01 support.
-		switch mediaType {
-		case "application/json", "text/json", "":
-			// ok
-		default:
-			httpFormatError(w, v, fmt.Errorf("unsupported media type: %q", mediaType))
-			return nil, false
-		}
-
-		// in v01 we actually get spans that we have to transform in traces
-		var spans []pb.Span
-		if err := json.NewDecoder(req.Body).Decode(&spans); err != nil {
-			log.Errorf("Cannot decode %s traces payload: %v", v, err)
-			httpDecodingError(err, []string{tagTraceHandler, fmt.Sprintf("v:%s", v)}, w)
-			return nil, false
-		}
-		traces = tracesFromSpans(spans)
-	case v02:
-		fallthrough
-	case v03:
-		fallthrough
-	case v04:
-		if err := decodeReceiverPayload(req.Body, &traces, v, mediaType); err != nil {
-			log.Errorf("Cannot decode %s traces payload: %v", v, err)
-			httpDecodingError(err, []string{tagTraceHandler, fmt.Sprintf("v:%s", v)}, w)
-			return nil, false
-		}
-	default:
-		httpEndpointNotSupported([]string{tagTraceHandler, fmt.Sprintf("v:%s", v)}, w)
-		return nil, false
-	}
-
-	return traces, true
-}
-
 func decodeReceiverPayload(r io.Reader, dest msgp.Decodable, v Version, mediaType string) error {
 	switch mediaType {
 	case "application/msgpack":
```
I feel this would work fine without all the weird functions.

Reply:
Good points. Yeah better not make this too complicated by returning errors and doing checking outside. 

Updated to move everything into `getTraces`.